[scroll-animations] Pass animation range value down into computeTimelineData

nmoucht opened the pull request:
#### 0aa547a211ad00f6c811fb509fd663b1c466cfba
<pre>
[scroll-animations] Pass animation range value down into computeTimelineData
<a href="https://bugs.webkit.org/show_bug.cgi?id=281543">https://bugs.webkit.org/show_bug.cgi?id=281543</a>
<a href="https://rdar.apple.com/138006146">rdar://138006146</a>

Reviewed by Tim Nguyen and Antoine Quint.

When calculating WebAnimation::currentTime, pass down any associated animation range down into
computeTimelineData. Also add parsing code for the JS api, reusing the parsing code for the CSS
portion. Also fix some various bugs about computing the timeline data using the animation range
value.

* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeTimelineData const):
(WebCore::ScrollTimeline::currentTime):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::currentTime):
(WebCore::ScrollTimeline::computeTimelineData):
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::defaultValue):
(WebCore::SingleTimelineRange::serialize const):
(WebCore::lengthFoCSSValue):
(WebCore::SingleTimelineRange::range):
(WebCore::SingleTimelineRange::parse):
(WebCore::TimelineRange::defaultForViewTimeline):
* Source/WebCore/animation/TimelineRange.h:
(WebCore::TimelineRange::isDefault const):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::computeTimelineData const):
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::currentTime const):
(WebCore::WebAnimation::setRangeStart):
(WebCore::WebAnimation::setRangeEnd):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::rangeStart const):
(WebCore::WebAnimation::rangeEnd const):
(WebCore::WebAnimation::setRangeStart): Deleted.
(WebCore::WebAnimation::setRangeEnd): Deleted.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnimationRangeStart):
(WebCore::Style::BuilderConverter::convertAnimationRangeEnd):
(WebCore::Style::BuilderConverter::convertAnimationRange): Deleted.

Canonical link: <a href="https://commits.webkit.org/285533@main">https://commits.webkit.org/285533@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c08f5b36447819d2a3e1fedfcb3dc5af8f1caeec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72902 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77108 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24136 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60132 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23952 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15813 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47305 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62748 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43951 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20231 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22473 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65807 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20572 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78773 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17148 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19716 "Build is being retried. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 20 flakes 3 failures; Uploaded test results; layout-tests running") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62754 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65040 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13359 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7016 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11225 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48125 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49192 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50487 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48937 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -52,7 +53,7 @@ class AnimationTimeline : public RefCounted<AnimationTimeline>, public CanMakeWe
     virtual void removeAnimation(WebAnimation&);
 
-    virtual std::optional<CSSNumberishTime> currentTime() { return m_currentTime; }
+    virtual std::optional<CSSNumberishTime> currentTime(TimelineRange = { }) { return m_currentTime; }
```
Should this be a `const` reference?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/DocumentTimeline.h:
```diff
@@ -58,7 +58,7 @@ class DocumentTimeline final : public AnimationTimeline
     Document* document() const { return m_document.get(); }
 
-    std::optional<CSSNumberishTime> currentTime() override;
+    std::optional<CSSNumberishTime> currentTime(TimelineRange = { }) override;
```
Same comment about using a `const` reference.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -169,10 +169,14 @@ float ScrollTimeline::floatValueForOffset(const Length& offset, float maxValue)
 }
 
-ScrollTimeline::Data ScrollTimeline::computeTimelineData(const TimelineRange& range) const
+ScrollTimeline::Data ScrollTimeline::computeTimelineData(const TimelineRange rangeIn) const
```
Could this remain a `const` reference?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,89 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+
+Length SingleTimelineRange::lengthFoCSSValue(RefPtr<const CSSPrimitiveValue> value)
+{
+    // TODO: Figure out how to make this work when provided calc value
```
@weinig may offer guidance here.

nmoucht replied:
I added the use of `CSSToLengthConversionData` which I think should resolve this.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,89 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+        return { };
+
+    // <timeline-range-name> <length-percentage>
+    auto& primitiveValue = downcast<CSSPrimitiveValue>(pair->second());
```
This yielded a safe C++ report, probably use a `Ref` here.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,89 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+            return range(*rangeValueOffset->toCSSValue(), type);
+    }
+
+    return { };
```
Given the number of `std::get_if` calls being made, it would be better to use `WTF::switchOn`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
```diff
@@ -383,7 +383,6 @@ JSValidityState.cpp
 JSVideoConfiguration.cpp
 JSVideoTrack.cpp
 JSVideoTrackList.cpp
-JSViewTimeline.cpp
```
I think this is unrelated to this specific patch, but rather belongs in https://github.com/WebKit/WebKit/pull/35312.

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,110 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+
+static const CSSToLengthConversionData cssToLengthConversionData(RefPtr<Element> element)
+{
+    auto* elementRenderer = element->renderer();
```
`CheckedPtr`

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,110 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+    auto* elementRenderer = element->renderer();
+    if (!elementRenderer)
+        return { };
+    auto* elementParentRenderer = elementRenderer->parent();
```
`CheckedPtr`

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,110 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+    if (!elementRenderer)
+        return { };
+    auto* elementParentRenderer = elementRenderer->parent();
+    auto& document = element->document();
```
Either use `Ref` or called `element->protectedDocument()`.

nmoucht replied:
Fixed.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
 FAIL Animation with animation-timeline:none holds current time at zero assert_equals: expected "100px" but got "0px"
-FAIL Animation with unknown timeline name holds current time at zero assert_equals: expected "100px" but got "101.9375px"
+FAIL Animation with unknown timeline name holds current time at zero assert_equals: expected "100px" but got "100.703125px"
```
No need to rebaseline that one.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        if (SingleTimelineRange::isOffsetValue(*primitiveValue)) {
+            // <length-percentage>
+            return { SingleTimelineRange::Name::Omitted, state ? Style::BuilderConverter::convertLength(*state, *primitiveValue) : lengthFoCSSValue(primitiveValue, element) };
```
This is just stylistic, but I'd put the comment above the `if` so I don't need braces.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+        tokenRange.consumeWhitespace();
+        if (auto consumedRange = type == SingleTimelineRange::Type::Start ? CSSPropertyParserHelpers::consumeAnimationRangeStart(tokenRange, parserContext) : CSSPropertyParserHelpers::consumeAnimationRangeEnd(tokenRange, parserContext))
+            return range(*consumedRange, type, nullptr, element);
+        return SingleTimelineRange { };
```
Might be wrong but you possibly don't need `SingleTimelineRange` here?

nmoucht replied:
This is necessary to fix build issues.

graouts commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+        }
+        if (RefPtr offset = rangeOffset.offset)
+            return range(*offset->toCSSValue(), type, nullptr, element);
+        return SingleTimelineRange { };
```
Same comment about not needing `SingleTimelineRange`.

nmoucht replied:
Ditto

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1834,4 +1834,17 @@ std::optional<double> WebAnimation::progress() const
 
+void WebAnimation::setRangeStart(TimelineRangeValue&& rangeStart)
+{
+    if (auto keyframeEffect = dynamicDowncast<KeyframeEffect>(m_effect.get()))
```
`RefPtr`

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1834,4 +1834,17 @@ std::optional<double> WebAnimation::progress() const
+
+void WebAnimation::setRangeEnd(TimelineRangeValue&& rangeEnd)
+{
+    if (auto keyframeEffect = dynamicDowncast<KeyframeEffect>(m_effect.get()))
```
`RefPtr`

nmoucht replied:
Fixed.

graouts reviewed (approved):
Patches like this ought to be split in smaller patches. This could for instance have had the `WebAnimation` changes to set the range start and end values be one patch, another one could have been plumbing `TimelineRange` values through the various APIs without making use of it, etc. It would make reviewing more focused.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+    if (!documentElement)
+        return std::nullopt;
+
+    return CSSToLengthConversionData { elementRenderer->style(), documentElement->renderer() ? &documentElement->renderer()->style() : nullptr, elementParentRenderer ? &elementParentRenderer->style() : nullptr, document->renderView() };
```
Would be easier to read like this:
```suggestion
    return CSSToLengthConversionData {
        elementRenderer->style(),
        documentElement->renderer() ? &documentElement->renderer()->style() : nullptr,
        elementParentRenderer ? &elementParentRenderer->style() : nullptr, 
        document->renderView()
    };
```

Also, do container query units work? It would be useful to pass the element for container query resolution (or add a FIXME to fix it)

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+    return CSSToLengthConversionData { elementRenderer->style(), documentElement->renderer() ? &documentElement->renderer()->style() : nullptr, elementParentRenderer ? &elementParentRenderer->style() : nullptr, document->renderView() };
+}
+
+Length SingleTimelineRange::lengthFoCSSValue(RefPtr<const CSSPrimitiveValue> value, RefPtr<Element> element)
```
```suggestion
Length SingleTimelineRange::lengthForCSSValue(RefPtr<const CSSPrimitiveValue> value, RefPtr<Element> element)
```

Typo ?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+        return value->resolveAsLength<WebCore::Length>(*conversionData);
+
+    if (value->isPercentage())
+        return WebCore::Length(value->resolveAsPercentage(*conversionData), LengthType::Percent);
```
nit: Drop the WebCore prefix from Length

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+        return WebCore::Length(value->resolveAsPercentage(*conversionData), LengthType::Percent);
+
+    if (value->isCalculatedPercentageWithLength())
+        return WebCore::Length(value->cssCalcValue()->createCalculationValue(*conversionData, CSSCalcSymbolTable { }));
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -111,4 +225,14 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const SingleTimelineRange& rang
 
+TimelineRange TimelineRange::defaultForScrollTimeline()
+{
+    return { { SingleTimelineRange::Name::Omitted, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::Start) }, { SingleTimelineRange::Name::Omitted, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::End) } };
```
Would be easier to read if you added new lines and indented.
```suggestion
    return {
        { SingleTimelineRange::Name::Omitted, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::Start) },
        { SingleTimelineRange::Name::Omitted, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::End) }
    };
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -111,4 +225,14 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const SingleTimelineRange& rang
+
+TimelineRange TimelineRange::defaultForViewTimeline()
+{
+    return { { SingleTimelineRange::Name::Cover, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::Start) }, { SingleTimelineRange::Name::Cover, SingleTimelineRange::defaultValue(SingleTimelineRange::Type::End) } };
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -305,28 +295,31 @@ ScrollTimeline::Data ViewTimeline::computeTimelineData(const TimelineRange& rang
     auto rangeEnd = computeRangeEnd() - insetStart.value();
 
-    return { currentScrollOffset, rangeStart + ScrollTimeline::floatValueForOffset(range.start.offset, rangeEnd - rangeStart), rangeEnd - ScrollTimeline::floatValueForOffset(range.end.offset, rangeEnd - rangeStart) };
+    return { currentScrollOffset, rangeStart + ScrollTimeline::floatValueForOffset(range.start.offset, rangeEnd - rangeStart), rangeStart + ScrollTimeline::floatValueForOffset(range.end.offset, rangeEnd - rangeStart) };
```
Could be easier to read split into multiple lines too

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -66,9 +67,10 @@ class ViewTimeline final : public ScrollTimeline {
 
 private:
-    ScrollTimeline::Data computeTimelineData(const TimelineRange& = { }) const override;
+    ScrollTimeline::Data computeTimelineData(const TimelineRange&) const override;
```
nit: those methods can probably be final

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -91,34 +91,18 @@ static std::optional<Length> lengthForInset(std::variant<RefPtr<CSSNumericValue>
-            return { lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedFirst())), lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedSecond())) };
-        return { lengthForInsetCSSValue(dynamicDowncast<CSSPrimitiveValue>(consumedInset)), std::nullopt };
+            return { SingleTimelineRange::lengthFoCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedFirst()), element), SingleTimelineRange::lengthFoCSSValue(dynamicDowncast<CSSPrimitiveValue>(insetPair->protectedSecond()), element) };
+        return { SingleTimelineRange::lengthFoCSSValue(dynamicDowncast<CSSPrimitiveValue>(consumedInset), element), std::nullopt };
```
Same here, I'd potentially try to split this into multiple lines.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,125 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
 
+Length SingleTimelineRange::defaultValue(Type type)
+{
+    return type == Type::Start ? Length(0, LengthType::Percent) : Length(100, LengthType::Percent);
```
The duplication between this and `percentIsDefault` would be nice to avoid.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/TimelineRange.cpp:
```diff
@@ -43,11 +48,125 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
+
+static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(RefPtr<Element> element)
+{
+    // FIXME: investigate container query units
```
```suggestion
    // FIXME: Investigate container query units.
```

also maybe put this before returning the `CSSToLengthConversionData`

nmoucht replied:
Fixed.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-none-expected.txt:
```diff
@@ -1,4 +1,4 @@
 
 FAIL Animation with animation-timeline:none holds current time at zero assert_equals: expected "100px" but got "0px"
-FAIL Animation with unknown timeline name holds current time at zero assert_equals: expected "100px" but got "101.9375px"
+FAIL Animation with unknown timeline name holds current time at zero assert_equals: expected "100px" but got "100.8125px"
```
No need to commit this flaky result change.

nmoucht committed:
[scroll-animations] Pass animation range value down into computeTimelineData
https://bugs.webkit.org/show_bug.cgi?id=281543
rdar://138006146

Reviewed by Tim Nguyen and Antoine Quint.

When calculating WebAnimation::currentTime, pass down any associated animation range down into
computeTimelineData. Also add parsing code for the JS api, reusing the parsing code for the CSS
portion. Also fix some various bugs about computing the timeline data using the animation range
value.

* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::currentTime):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::computeTimelineData const):
(WebCore::ScrollTimeline::currentTime):
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::currentTime):
(WebCore::ScrollTimeline::computeTimelineData):
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::defaultValue):
(WebCore::SingleTimelineRange::serialize const):
(WebCore::lengthFoCSSValue):
(WebCore::SingleTimelineRange::range):
(WebCore::SingleTimelineRange::parse):
(WebCore::TimelineRange::defaultForViewTimeline):
* Source/WebCore/animation/TimelineRange.h:
(WebCore::TimelineRange::isDefault const):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::computeTimelineData const):
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::currentTime const):
(WebCore::WebAnimation::setRangeStart):
(WebCore::WebAnimation::setRangeEnd):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::rangeStart const):
(WebCore::WebAnimation::rangeEnd const):
(WebCore::WebAnimation::setRangeStart): Deleted.
(WebCore::WebAnimation::setRangeEnd): Deleted.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAnimationRangeStart):
(WebCore::Style::BuilderConverter::convertAnimationRangeEnd):
(WebCore::Style::BuilderConverter::convertAnimationRange): Deleted.

Canonical link: https://commits.webkit.org/285533@main
